### PR TITLE
gspo support

### PR DESCRIPTION
**Pull request by Kenwwww**
add gspo loss support

### CODE REVIEW

File: verl/workers/fsdp_workers.py [hunk 1]
```diff
@@ -1,699 +1,700 @@
...
+        if self._use_param_offload:
+            load_fsdp_model(self.fsdp_module)
+
+        if self._use_optimizer_offload:
+            load_fsdp_optimizer(optimizer=self.optimizer)
+
+        with self.ulysses_sharding_manager:
+            data = self.ulysses_sharding_manager.preprocess_data(data=data)
+            with Timer(name="update_policy", logger=None) as timer:
+                metrics = self.actor.update_policy(data=data)
```
[the remark this answers is not in the document]
Comment (Kenwwww) [reply verl/workers/fsdp_workers.py]: fix it

File: verl/trainer/core_algos.py [hunk 2]
```diff
@@ -1,519 +1,624 @@
...
+            torch.log(torch.tensor(1.0 - clip_ratio_low, device=log_ratio.device)),
+            torch.log(torch.tensor(1.0 + clip_ratio_high, device=log_ratio.device)),
+        )
+    )
+    metrics = {}
+
+    pg_loss = -advantages * ratio  # shape: [batch_size, seq_len]
+    pg_loss2 = -advantages * clipped_ratio
+
+    clip_ratio_dual_val = 1.0 - clip_ratio_dual
```
Comment (TristanMeng) [inline verl/trainer/core_algos.py:600]: hello, i find there are some problem. clamp will make 'nan'. so there code is what funtion?

File: verl/trainer/core_algos.py [hunk 3]
```diff
@@ -1,519 +1,624 @@
...
+        return torch.clamp(kld, min=-10.0, max=10.0)
+
+    if kl_penalty == "full":
+        return F.kl_div(ref_log_probs, log_probs, log_target=True, reduction="none").sum(-1)
+
+    raise NotImplementedError(f"Unknown KL penalty: {kl_penalty}.")
+
+
+# gspo loss
+def compute_policy_loss_gspo(
```
[the remark this answers is not in the document]
Comment (TristanMeng) [reply verl/trainer/core_algos.py:568]: this is my code ，review, thanks.

def compute_policy_loss_gspo(
    old_log_probs: torch.Tensor,
    log_probs: torch.Tensor,
    advantages: torch.Tensor,
    response_mask: torch.Tensor,
    clip_ratio_low: float,
    clip_ratio_high: float,
    clip_ratio_dual: float,
    loss_avg_mode: Literal["token", "seq"],
) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
    negative_approx_kl = log_probs - old_log_probs  # [batch_size, seq_len]

    # compute sequence-level importance ratio:
    # si(θ) = (π_θ(yi|x)/π_θold(yi|x))^(1/|yi|) =
    # exp [(1/|y_i|) * Σ_t log(π_θ(y_i,t|x,y_i,<t)/π_θold(y_i,t|x,y_i,<t))]
    seq_lengths = torch.sum(response_mask, dim=-1).clamp(min=1)
    negative_approx_kl_seq = torch.sum(negative_approx_kl * response_mask, dim=-1) / seq_lengths

    # Combined ratio at token level:
    # s_i,t(θ) = sg[s_i(θ)] · π_θ(y_i,t|x, y_i,<t) / sg[π_θ(y_i,t|x, y_i,<t)]
    # In log space: log(s_i,t(θ)) = sg[log(s_i(θ))] + log_prob - sg[log_prob]
    log_seq_importance_ratio = log_probs - log_probs.detach() + negative_approx_kl_seq.detach().unsqueeze(-1)
    log_seq_importance_ratio = torch.clamp(log_seq_importance_ratio, max=10.0)  # clamp for numerical stability

    # finaly exp() to remove log
    seq_importance_ratio = torch.exp(log_seq_importance_ratio)

    # pg metrics
    metrics = {"ppo_kl": -negative_approx_kl}
    # use negative log probs as an estimator of entropy loss
    metrics["entropy_loss"] = average_loss(-log_probs, response_mask, mode=loss_avg_mode)

    pg_losses1 = -advantages * seq_importance_ratio #-ratio * A
    pg_losses2 = -advantages * torch.clamp(seq_importance_ratio, 1 - clip_ratio_low, 1 + clip_ratio_high) #-clip(ratio, 1-clip_low, 1+clip_high) * A
    pg_losses = torch.maximum(pg_losses1, pg_losses2)

    metrics["pg_clipfrac_higher"] = (pg_losses1 < pg_losses2).float()
    # clipped_pg_loss_lower = torch.min(clipped_pg_loss_higher, pg_loss3)  # clip if pg_loss > pg_loss3 and adv < 0
    # final_pg_loss = torch.where(advantages < 0, clipped_pg_loss_lower, clipped_pg_loss_higher)
    # metrics["pg_clipfrac_lower"] = (clipped_pg_loss_higher > pg_loss3).float() * (advantages < 0).float()

    metrics = {k: VF.masked_mean(v, response_mask).detach().item() for k, v in metrics.items()}
    # for GSPO, we need to aggregate the loss at the sequence level (seq-mean-token-mean)
    final_pg_loss = average_loss(pg_losses, response_mask, mode="seq")

    return final_pg_loss, metrics

[the remark this answers is not in the document]
Comment (TristanMeng) [reply verl/trainer/core_algos.py:568]: should use Dual-clip ?

**Comment by SStoica12** [comment]
@Kenwwww and @TristanMeng Can I ask about progress on this? I am also interested in seeing gspo with EasyR1.

**Comment by TristanMeng** [comment]
> @Kenwwww and @TristanMeng Can I ask about progress on this? I am also interested in seeing gspo with EasyR1.

ok, I can't improve reward using grpo on 30B-A3B. even 5 epoch. but when using gspo, after 3 epoch, reward will greatly improve.
